Tuple Data Collection

lnkuiper opened the pull request:
This PR implements `TupleDataCollection`, an (even) more unified way of materializing data in row format.

In various places in DuckDB's execution engine, we materialize data in row format using `RowDataCollection` and `RowLayout`. While this is already a somewhat unified way of materializing data in row format, it has a few issues when we want to spill the row data to disk. The pointers within the row format have to be manually (un)swizzled when (un)pinned. This is cumbersome and requires some serious bookkeeping within physical operators that use row format data. This (un)swizzling may also be redundant if the blocks that hold the data are never written to disk, which may happen in cases where the data size is only slightly over the memory limit.

This PR implements the `TupleDataCollection` class that, similar to the `ColumnDataCollection` class (#4315), `DataChunk`s can simply be appended or scanned, without the need to explicitly hold buffer pins in various places in the code. This is all managed by the `TupleDataAllocator` class. The two problems described above (regarding pointer swizzling) are addressed in a similar way to how pointers in a `ColumnDataCollection` are never explicitly swizzled (#5543). In short, we simply compare pointers to see whether the block was spilled to disk. If the pointers are different, we recompute them, if the pointers are the same, they are still valid and nothing needs to be done.

`TupleDataCollection` has been integrated in `JoinHashTable` and `AggregateHashTable`, and should be integrated into the sorting code before the old `RowDataCollection` can be phased out. At that point I would like to rename `TupleDataCollection` to `RowDataCollection` for more consistency with naming in the code base.

This PR also brings improvements to the out-of-core hash join, and should improve the out-of-core capabilities of the hash aggregate. To improve the out-of-core hash join, I've implemented `PartitionedTupleData`, which, similar to `PartitionedColumnData` (#4970), we can append `DataChunk`s to, directly to the partitions. Rather than materializing all data in a single `TupleDataCollection`, we now immediately materialize to 16 partitions. If all data fits in memory, we combine the partitions into one, and perform an in-memory hash join. If not all data fits in memory, we combine as many partitions as fit in memory into one, and do a partitioned hash join like before. If a single one of the 16 partitions does not fit in memory, we repartition to up to 256 partitions, and perform the out-of-core hash join like before.

I've ran the same benchmark as in the previous PRs (linked above):

| Memory limit (GB) | Old time (s) | New time (s) |
|-------------------|--------------|--------------|
| 10                | 1.74         | 1.56         |
| 5                 | 2.47         | 2.09         |
| 3                 | 4.54         | 3.64         |
| 1                 | 4.80         | 3.66         |

I got a bit side-tracked in this PR and vectorized some code in `hive_partitioning.cpp` and was able to improve the runtime of `to_parquet_partition_by_many.benchmark` from ~2.95s to ~2.25s on my laptop.

P.S.: I am definitely not done tweaking the performance of the hash join/aggregate, especially after speaking with Thomas Neumann here at ICDE. I look forward to optimizing these operators more in the future.

Also, this definitely has merge conflicts with #6995. Happy to fix them and merge this after that has been merged.

Commit:
merge with master

Commit:
refactor folder structure so row/column stuff is together

Commit:
init TupleDataCollection

Commit:
progress

Commit:
big ol' merge with master after the christmas holidays

Commit:
merge with master and TupleDataCollection progress

Commit:
more tuple data collection progress

Commit:
TupleData segment/chunks and progress on inline struct scatter/gather

Commit:
some list gather/scatter - not finished yet

Commit:
Merge branch 'master' into oochj

Commit:
Merge branch 'master' into oochj

Commit:
implement TupleDataCollection append and scan

Commit:
Merge branch 'master' into oochj

Commit:
use 4byte offset instead of 8byte pointer to save space in TupleDataLayout

Commit:
add selection vector to scatter/gather

Commit:
implement ComputeHeapSizes

Commit:
add way to keep everything pinned

Commit:
Merge branch 'master' into oochj

Commit:
implement PartitionedTupleData

Commit:
just use the layout in the constructor

Commit:
integrate TupleDataCollection in PhysicalHashJoin (compiling again yay)

Commit:
Merge branch 'master' into oochj

Commit:
actually running some joins again

Commit:
remove (almost) duplicate code - more code re-use yay

Commit:
make sure KEEP_EVERYTHING_PINNED works correctly

Commit:
Merge branch 'master' into oochj

Commit:
unify FinalizeChunkState

Commit:
add chunk part merging

Commit:
implement new struct match

Commit:
store heap size in each row instead of offset - also get RowLayout from TupleDataLayout for compatibility (for now)

Commit:
minor fixes and a bunch of comments to explain the API

Commit:
unify TupleDataCollection Scatter/Gather with RowOperations

Commit:
Merge branch 'master' into oochj

Commit:
separate pin/chunk state so for eventual tuple data partitioning

Commit:
re-write chunk part merging for clarity

Commit:
split up append into build/scatter so PartitionedTupleData is more efficient

Commit:
implement re-partitioning of PartitionedTupleData and integrate into hash join

Commit:
Merge branch 'master' into oochj

Commit:
implement appending rows to TupleDataCollection

Commit:
debug some string issues

Commit:
debug more string stuff

Commit:
external join with many rounds working again

Commit:
merge with master

Commit:
big external join working again

Commit:
more bugfixes and verification

Commit:
fix issue where empty heap would segfault

Commit:
fix issue with tuple data iterator not keeping blocks pinned

Commit:
set tests back to normal after messing with them for a bit

Commit:
Merge branch 'master' into oochj

Commit:
remove old row partitioning code

Commit:
store partition entries in state instead of re-initializing every time

Commit:
optimize row validity mask initialization

Commit:
getting the perfect hash join to work again

Commit:
perfect hash join working again

Commit:
first draft of new list row serialization

Commit:
basic list row serialization working

Commit:
merge with master

Commit:
lots of progress getting test_nested_payloads.test to work

Commit:
nested payloads (mostly) working

Commit:
Merge branch 'master' into oochj

Commit:
make sure we orrify exactly once

Commit:
improve nested type serialization and move scatter/gather functions to separate file

Commit:
get rid of TupleDataScanProperties and use TupleDataCollection in row_match.cpp

Commit:
get rid of unused param

Commit:
Merge branch 'master' into oochj

Commit:
fix list gather for row_match

Commit:
integrate TupleDataCollection into GroupedAggregateHashTable

Commit:
successfully running group by's on TupleDataCollection :D

Commit:
all tests in test_group_by.test working

Commit:
get rid of PayloadApply in favour of TupleDataChunkIterator

Commit:
merge with master

Commit:
remove more recursive cte checks

Commit:
unify radix partitioning stuff and debug full/outer joins

Commit:
all fast tests working now with TupleDataCollection

Commit:
merge with master

Commit:
fix issue where pins were held in random order

Commit:
add missing includes

Commit:
get rid of partition buffer in PartitionedColumnData (performance) and fix issue with held tuple data pins

Commit:
optimize some loops

Commit:
fix another bug with held pins

Commit:
forward declaration for struct layouts

Commit:
wrap struct layouts in a unique_ptr so gcc does not complain

Commit:
Merge branch 'master' into oochj

Commit:
remove wrong verification

Commit:
format on linux box so CI is happy

Commit:
fix bug with null BLOBs

Commit:
full include paths so amalgamation.py can find the files

Commit:
vectorize hive key partition index computation

Commit:
Merge branch 'master' into oochj

Commit:
Merge branch 'oochj' of github.com:lnkuiper/duckdb into oochj

Commit:
some CodeQuality fixes and don't recompute heap pointers if NULL

Commit:
properly apply selection vector when dealing with structs within nested lists

Commit:
make sure we don't string verify AGGREGATE_STATE

Commit:
Merge branch 'master' into oochj

Commit:
don't destroy heap blocks too early

Commit:
implement data unpinning in Finalize of agg HT

Commit:
slightly improve partitioned append performance

Commit:
merge include with master

Commit:
merge with master after all the make_uniq shenanigans

Commit:
merge with master after cast changes

Commit:
properly unpin stuff

Commit:
properly fix symbols after make_uniq stuff

Commit:
don't set swizzling if we aren't reordering / swizzling heap

Commit:
Merge branch 'master' into oochj

Commit:
merge with master

Commit:
make aggregate_allocator shared because partition does not copy over the data anymore

lnkuiper commented:
Just merged with master after the #6995 changes, and I found an issue with the `aggregate_allocator` and the new partitioning code. The new partitioning code copies whole rows to the partitions without scanning, appending, and combining aggregate states. This is much more efficient than the previous implementation.

For the list aggregate, the unpartitioned HT, which has the original `aggregate_allocator`, is still the owner of the data in the aggregate states of the partitioned HTs. When the unpartitioned HT gets destroyed we get an assertion failure that there are still outstanding allocations that have not been de-allocated. This is correct, but these allocations still need to stay around.

My solution is to change `aggregate_allocator` to a `shared_ptr<ArenaAllocator>`, and make sure the partitions have the same allocator when we partition.

Although `ArenaAllocator` is not thread-safe, this should work fine because the partitions are only used by a single thread at a time when adding chunks into them.

Reading the aggregate HTs in parallel should also not be a problem during the combine, as long as the HT that we sink into is "fresh", i.e., empty and without an allocator, which I think it is.

What do you think @Mytherin ?

Mytherin commented:
> Just merged with master after the #6995 changes, and I found an issue with the `aggregate_allocator` and the new partitioning code. The new partitioning code copies whole rows to the partitions without scanning, appending, and combining aggregate states. This is much more efficient than the previous implementation.
> 
> For the list aggregate, the unpartitioned HT, which has the original `aggregate_allocator`, is still the owner of the data in the aggregate states of the partitioned HTs. When the unpartitioned HT gets destroyed we get an assertion failure that there are still outstanding allocations that have not been de-allocated. This is correct, but these allocations still need to stay around.
> 
> My solution is to change `aggregate_allocator` to a `shared_ptr<ArenaAllocator>`, and make sure the partitions have the same allocator when we partition.
> 
> Although `ArenaAllocator` is not thread-safe, this should work fine because the partitions are only used by a single thread at a time when adding chunks into them.
> 
> Reading the aggregate HTs in parallel should also not be a problem during the combine, as long as the HT that we sink into is "fresh", i.e., empty and without an allocator, which I think it is.
> 
> What do you think @Mytherin ?

I wonder if the `ArenaAllocator::Move` method cannot be used instead to move over the allocator data from one hash table to another?

lnkuiper commented:
@Mytherin The problem is that the arena allocator contains the list aggregate data for the whole unpartitioned HT, but when we partition, the list data is spread across all partitions, but there is only one arena allocator that owns the data. We could move the arena allocator somewhere else, perhaps? It should be somewhere where it stays alive until the aggregate is complete.

Mytherin commented:
> @Mytherin The problem is that the arena allocator contains the list aggregate data for the whole unpartitioned HT, but when we partition, the list data is spread across all partitions, but there is only one arena allocator that owns the data. We could move the arena allocator somewhere else, perhaps? It should be somewhere where it stays alive until the aggregate is complete.

Aha, that makes sense. I think that is indeed fine as long as the partitions are all thread-local (which I think they should be).

lnkuiper commented:
Thinking about this more, maybe we can have both an `aggregate_allocator` which can be written to (non-shared) and a shared one that is just there to keep the data alive. That is much less error-prone. I can add a lock so we don't run into issues when multiple threads are repartitioning in parallel.

Mytherin commented:
> Thinking about this more, maybe we can have both an `aggregate_allocator` which can be written to (non-shared) and a shared one that is just there to keep the data alive. That is much less error-prone. I can add a lock so we don't run into issues when multiple threads are repartitioning in parallel.

The problem is that freeing data will get messy if allocators are mixed - although freeing data individually is obviously not required for an arena allocator. An alternative could also be that we deprecate the `destroy` callback for the aggregate function entirely, and force everything to be allocated through arena allocators that are passed in from the hash table itself. That might make the most sense - but it would require some work with migrating our existing aggregate functions to all use the arena allocator infrastructure.

Commit:
store pre-partitioning allocators

Commit:
Merge branch 'master' into oochj

Commit:
lock chunk parts when recomputing pointers

Commit:
check before grabbing lock

Commit:
get rid of vector of allocators again

hawkfish commented on src/include/duckdb/common/types/row/tuple_data_segment.hpp:
```diff
@@ -0,0 +1,122 @@
+	uint32_t heap_block_index;
+	uint32_t heap_block_offset;
+	data_ptr_t base_heap_ptr;
+	//! Size of the last heap row (for all but the last we can infer the size from the pointer difference)
```
The comment suggests this is the entire heap, not just one row?

lnkuiper replied:
Thanks, good catch! This comment is outdated. This now stores the heap size for this chunk part.

hawkfish commented on src/include/duckdb/common/types/row/tuple_data_allocator.hpp:
```diff
@@ -0,0 +1,116 @@
+
+public:
+	//! Remaining capacity (in bytes)
+	idx_t RemainingCapacity() {
```
const? Or is the point that it is not thread safe?

hawkfish commented on src/include/duckdb/common/types/row/tuple_data_allocator.hpp:
```diff
@@ -0,0 +1,116 @@
+class TupleDataAllocator {
+public:
+	TupleDataAllocator(BufferManager &buffer_manager, const TupleDataLayout &layout);
+	TupleDataAllocator(TupleDataAllocator &allocator);
```
explicit

hawkfish commented on src/include/duckdb/common/types/row/tuple_data_collection.hpp:
```diff
@@ -0,0 +1,235 @@
+	//! Unpins all held pins
+	void Unpin();
+
+	//! Gets the scatter function for the given column index
```
s/index/type/?

lnkuiper replied:
Thanks, this comment was also outdated.

hawkfish commented on src/include/duckdb/common/types/row/tuple_data_collection.hpp:
```diff
@@ -0,0 +1,235 @@
+	            const idx_t append_count = DConstants::INVALID_INDEX);
+	//! Append a DataChunk to this TupleDataCollection using the specified pin and chunk states
+	//! - ToUnifiedFormat has already been called
+	void AppendUnified(TupleDataPinState &pin_state, TupleDataChunkState &chunk_state, DataChunk &new_chunk,
```
Is the argument order here a legacy of renaming the method?

hawkfish commented on src/include/duckdb/common/types/row/tuple_data_collection.hpp:
```diff
@@ -0,0 +1,235 @@
+	//! Builds out the buffer space for the specified chunk state
+	void Build(TupleDataPinState &pin_state, TupleDataChunkState &chunk_state, const idx_t append_offset,
+	           const idx_t append_count);
+	//! Scatters the given DataChunk to the rows in the specified append state
```
Is there an implied append state?

lnkuiper replied:
Good catch, this comment is also outdated. An append state is nothing more than:
```c++
struct TupleDataAppendState {
	TupleDataPinState pin_state;
	TupleDataChunkState chunk_state;
};
```
However, in many cases it was useful to pull these apart, and supply the pin/chunk states separately. For example, when appending to a `PartitionedTupleDataCollection`, we need one pin state per partition, but only one chunk state overall.

hawkfish commented on src/include/duckdb/common/types/row/tuple_data_collection.hpp:
```diff
@@ -0,0 +1,235 @@
+	void FinalizePinState(TupleDataPinState &pin_state);
+
+	//! Appends the other TupleDataCollection to this, destroying the other data collection
+	void Combine(TupleDataCollection &other);
```
Could this be a move operation?

hawkfish commented on src/include/duckdb/common/types/row/tuple_data_iterator.hpp:
```diff
@@ -0,0 +1,64 @@
+	//! Resets the scan indices to the start
+	void Reset();
+	//! Get the count of the current "DataChunk"
+	idx_t GetCount() const;
```
GetCurrentChunkCount?

hawkfish commented on src/include/duckdb/execution/join_hashtable.hpp:
```diff
@@ -222,13 +215,12 @@ class JoinHashTable {
-	vector<BufferHandle> pinned_handles;
+	//! TODO
+	mutex data_lock;
+	//! TODO
```
Can this be commented not?

hawkfish reviewed (commented):
Looks nice!  I've only looked at the header files because you asked me to comment on the API. It looks much better than the old stuff - most of my comments are more about clarifying the comments/documentation.

lnkuiper commented on src/include/duckdb/common/types/validity_mask.hpp:
```diff
@@ -170,6 +170,9 @@ struct TemplatedValidityMask {
 		}
 		return ValidityBuffer::MAX_ENTRY >> (BITS_PER_VALUE - n);
 	}
+	static inline idx_t SizeInBytes(idx_t n) {
```
No, this is the size in bytes:
```
  (n + BITS_PER_VALUE - 1) / BITS_PER_VALUE
```
Let's say `BITS_PER_VALUE` is 8, and we supply `n = 3`, then we will get:
```
  (3 + 8 - 1) / 8
= 12 / 8
= 1
```
Gives you the size of a validity mask that can hold  `n` values.

lnkuiper commented on src/include/duckdb/common/radix_partitioning.hpp:
```diff
@@ -9,52 +9,61 @@
-	static constexpr const idx_t NUM_RADIX_BITS = radix_bits;
-	static constexpr const idx_t NUM_PARTITIONS = (idx_t)1 << NUM_RADIX_BITS;
-	static constexpr const idx_t TMP_BUF_SIZE = 8;
+	static inline constexpr idx_t NumberOfPartitions(const idx_t &radix_bits) {
```
I've gotten used to using a const reference everywhere, because clang-tidy will complain about this when the arguments to a function are not trivially copyable, like `vector`. Of course, `idx_t` is trivially copyable so this is not necessary, but it does not hurt, either.

lnkuiper commented on src/include/duckdb/common/radix_partitioning.hpp:
```diff
@@ -9,52 +9,61 @@
-		return (hash & MASK) >> (sizeof(hash_t) * 8 - NUM_RADIX_BITS);
+	static inline idx_t RadixBits(const idx_t &n_partitions) {
+		D_ASSERT(IsPowerOfTwo(n_partitions));
+		for (idx_t r = 0; r < sizeof(idx_t) * 8; r++) {
```
It's the inverse of `NumberOfPartitions`, I've clarified it with a comment

lnkuiper commented on src/include/duckdb/common/radix_partitioning.hpp:
```diff
@@ -101,4 +111,45 @@ class RadixPartitionedColumnData : public PartitionedColumnData {
+	void InitializeAppendStateInternal(PartitionedTupleDataAppendState &state,
+	                                   TupleDataPinProperties properties) const override;
+	void ComputePartitionIndices(PartitionedTupleDataAppendState &state, DataChunk &input) override;
+	void ComputePartitionIndices(Vector &row_locations, idx_t count, Vector &partition_indices) const override;
```
Maybe the code can be shared? I have some code duplication `PartitionedColumnData` and `PartitionedTupleData`, which I want to get rid of in the future.

lnkuiper commented on src/include/duckdb/common/types/row/tuple_data_layout.hpp:
```diff
@@ -0,0 +1,113 @@
+	TupleDataLayout Copy() const;
+
+public:
+	//! Initializes the RowLayout with the specified types and aggregates to an empty RowLayout
```
They're essentially the same, but `TupleDataLayout` inlines structs within the layout rather than storing a pointer and writing them to the heap.

This is temporary, of course. I want to phase out the old `Row*` stuff and replace it with `TupleData*` stuff, but this PR is big enough. Once everything is replaced, we can rename it back, but for now, both need to exist.

lnkuiper commented on src/include/duckdb/common/types/row/tuple_data_states.hpp:
```diff
@@ -0,0 +1,74 @@
+};
+
+struct TupleDataPinState {
+	unordered_map<uint32_t, BufferHandle> row_handles;
```
It's block id indeed, same as in `column_data_scan_states.hpp` (although I think an `idx_t` is used there).

lnkuiper commented on src/include/duckdb/common/types/row/tuple_data_collection.hpp:
```diff
@@ -0,0 +1,235 @@
+	//! Gets the UnifiedVectorFormat from the chunk state as an array
+	static unique_ptr<UnifiedVectorFormat[]> GetVectorData(TupleDataChunkState &chunk_state);
+	//! Computes the heap sizes for the new DataChunk that will be appended
+	static void ComputeHeapSizes(TupleDataChunkState &chunk_state, DataChunk &new_chunk,
```
Yes, thanks. I've added a bunch more const to `DataChunk`s and `Vector`s in the append/scatter code.

Commit:
clean up tuple data API (Richard's PR feedback)

Commit:
Merge branch 'master' into oochj

lnkuiper commented:
Hi @hawkfish, big thanks for the review. I've cleaned up some comments, added `const` here and there, and improved function names. Hopefully, that makes future usage easier. Just looking to clear up the final thread sanitizer issue now.

Commit:
slightly higher memory limit so we don't go over by chance (parallelism)

Mytherin commented on src/include/duckdb/common/types/row/tuple_data_collection.hpp:
```diff
@@ -0,0 +1,237 @@
+	//! Creates a UnifiedVectorFormat in the given Chunk state for the given DataChunk
+	static void ToUnifiedFormat(TupleDataChunkState &chunk_state, DataChunk &new_chunk);
+	//! Gets the UnifiedVectorFormat from the Chunk state as an array
+	static unique_ptr<UnifiedVectorFormat[]> GetVectorData(const TupleDataChunkState &chunk_state);
```
This forces the `UnifiedVectorFormat[]` to be allocated for each call - likely not very significant but we could take it as a parameter instead and cache it (which is what we do in the `AggregateHashtable` now).

Mytherin commented on src/include/duckdb/execution/aggregate_hashtable.hpp:
```diff
@@ -139,24 +141,23 @@ class GroupedAggregateHashTable : public BaseAggregateHashTable {
 	//! The data of the HT
-	vector<BufferHandle> payload_hds;
+	unique_ptr<TupleDataCollection> data_collection;
+	TupleDataAppendState td_append_state;
```
Should this be moved to the `AggregateHTAppendState`?

lnkuiper replied:
Good point, I will try to do this. I can move the `TupleDataChunkState` of the `TupleDataAppendState` into the `AggregateHTAppendState`, but the `TupleDataPinState` of the `TupleDataAppendState` needs to be HT-specific.

Mytherin commented on src/include/duckdb/common/types/row/partitioned_tuple_data.hpp:
```diff
@@ -0,0 +1,140 @@
+	SelectionVector partition_sel;
+
+	static constexpr idx_t MAP_THRESHOLD = 32;
+	perfect_map_t<list_entry_t> partition_entries;
```
Is a `perfect_map` not better implemented as a `vector`, or are the numbers here potentially very large?

lnkuiper replied:
The usage so far almost always has small numbers but can in some cases have large numbers, but the size of the map (number of entries) will always be small. Therefore I chose not to use a vector. I want to get back to this at some point and unify computing the partition indices for both the `PartitionedColumnData` and `PartitionedTupleData`, and then I will try to find a more elegant solution.

Mytherin reviewed (commented):
Thanks for the PR! Looks great. Very nice results, and great new API! Some comments below:

Commit:
Merge branch 'master' into oochj

lnkuiper commented on src/common/hive_partitioning.cpp:
```diff
@@ -140,29 +141,157 @@ void HivePartitioning::ApplyFiltersToFileList(ClientContext &context, vector<str
+			auto &key = keys[i];
+			const auto idx = sel.get_index(i);
+			if (validity.RowIsValid(idx)) {
+				key.values[col_idx] = GetHiveKeyValue(data[idx], type);
```
Yes, it is faster. We need `Value` for this to be generic, but there's possibly other ways of achieving this.
I didn't want to touch the implementation too much, so I did it like this.

The problem with the initial implementation was that it called `GetValue`  with a col/row idx on the `DataChunk` to get the values, which is a pretty expensive function to call multiple times. My rework vectorizes getting the data from the vector, which is significantly faster.

As I said in my PR:
> I got a bit side-tracked in this PR and vectorized some code in hive_partitioning.cpp and was able to improve the runtime of to_parquet_partition_by_many.benchmark from ~2.95s to ~2.25s on my laptop.

Commit:
trying to fix race condition and some PR feedback

Commit:
more PR feedback

Commit:
fix include path

lnkuiper commented:
Thanks for the feedback, Mark :)

Just trying to please CI now

Commit:
Merge branch 'master' into oochj

Commit:
grab lock in destructor so threadsan does not complain

Commit:
merge with master and unpin when done (oops)

Commit:
merge with master

Commit:
Merge branch 'master' into oochj

lnkuiper commented:
I'm getting data races in the buffer eviction queue, and at this point I'm not sure what to do about it anymore. I'm also unable to reproduce this on a linux box

Commit:
suppress AddToEvictionQueue

Mytherin commented:
Thanks! LGTM
